app/lib/search/index_simple.dart

```diff
@@ -71,66 +71,59 @@ class SimplePackageIndex implements PackageIndex {
 
   @override
   Future<PackageSearchResult> search(SearchQuery query) async {
-    final Map<String, double> total = <String, double>{};
-    void addAll(Map<String, double> scores, double weight) {
-      scores?.forEach((String url, double score) {
-        if (score != null) {
-          final double prev = total[url] ?? 0.0;
-          total[url] = prev + score * weight;
-        }
-      });
-    }
-
-    addAll(_nameIndex.search(query.text), 0.70);
-    addAll(_descrIndex.search(query.text), 0.10);
-    addAll(_readmeIndex.search(query.text), 0.05);
-
-    if ((query.text == null || query.text.isEmpty) &&
-        query.packagePrefix != null) {
-      addAll(_nameIndex.search(query.packagePrefix), 0.8);
+    // do text matching
+    final Score textScore = _searchText(query.text, query.packagePrefix);
+
+    // The set of urls to filter on.
+    final Set<String> urls =
+        textScore?.getKeys()?.toSet() ?? _documents.keys.toSet();
+
+    // filter on package prefix
+    if (query.packagePrefix != null) {
+      urls.removeWhere(
+        (url) => !_documents[url]
+            .package
+            .toLowerCase()
+            .startsWith(query.packagePrefix.toLowerCase()),
+      );
     }
 
-    addAll(getHealthScore(total.keys), 0.05);
-    addAll(getPopularityScore(total.keys), 0.10);
-
-    List<PackageScore> results = <PackageScore>[];
-    for (String url in total.keys) {
-      final PackageDocument doc = _documents[url];
-
-      // filter on platform
-      if (query.platformPredicate != null &&
-          !query.platformPredicate.matches(doc.platforms)) {
-        continue;
-      }
-
-      // filter on package prefix
-      if (query.packagePrefix != null &&
-          !doc.package
-              .toLowerCase()
-              .startsWith(query.packagePrefix.toLowerCase())) {
-        continue;
-      }
-
-      results.add(new PackageScore(
-        url: doc.url,
-        package: doc.package,
-        score: total[url],
-      ));
+    // filter on platform
+    if (query.platformPredicate != null) {
+      urls.removeWhere(
+          (url) => !query.platformPredicate.matches(_documents[url].platforms));
     }
 
-    results.sort((a, b) => -a.score.compareTo(b.score));
-
-    // filter out the noise (maybe a single matching ngram)
-    if (results.isNotEmpty) {
-      final double bestScore = results.first.score;
-      final double scoreTreshold = bestScore / 25;
-      results.removeWhere((pr) => pr.score < scoreTreshold);
+    // reduce text results if filter did remove an url
+    textScore?.removeWhere((key) => !urls.contains(key));
+
+    List<PackageScore> results;
+    switch (query.order ?? SearchOrder.overall) {
+      case SearchOrder.overall:
+        final Score overallScore = new Score()
+          ..addValues(textScore?.values, 0.85)
+          ..addValues(getPopularityScore(urls), 0.10)
+          ..addValues(getHealthScore(urls), 0.05);
+        results = _flattenFromValues(overallScore.values);
```

Comment:
Instead of 'flatten' we could use 'rank', because what it really is doing is ordering/ranking the results.